discojs/src/training/trainer.ts

```diff
@@ -0,0 +1,142 @@
+import * as tf from "@tensorflow/tfjs";
+import { List } from "immutable";
+
+import type {
+  BatchLogs,
+  EpochLogs,
+  Model,
+  Task,
+  WeightsContainer,
+} from "../index.js";
+import { privacy } from "../index.js";
+import { Client } from "../client/index.js";
+import * as async_iterator from "../utils/async_iterator.js";
+
+export interface RoundLogs {
+  epochs: List<EpochLogs>;
+  participants: number;
+}
+
+/** Train a model and exchange with others **/
+export class Trainer {
+  readonly #client: Client;
+  readonly #roundDuration: number;
+  readonly #epochs: number;
+  readonly #privacy: Task["trainingInformation"]["privacy"];
+
+  #training?: AsyncGenerator<
+    AsyncGenerator<AsyncGenerator<BatchLogs, EpochLogs>, RoundLogs>,
+    void
+  >;
+
+  constructor(
+    task: Task,
+    public readonly model: Model,
+    client: Client,
+  ) {
+    this.#client = client;
+    this.#roundDuration = task.trainingInformation.roundDuration;
+    this.#epochs = task.trainingInformation.epochs;
+    this.#privacy = task.trainingInformation.privacy;
+
+    if (!Number.isInteger(this.#epochs / this.#roundDuration))
+      throw new Error(
+        `round duration ${this.#roundDuration} doesn't divide number of epochs ${this.#epochs}`,
+      );
+  }
+
+  async stopTraining(): Promise<void> {
+    await this.#training?.return();
+  }
+
+  async *train(
+    dataset: tf.data.Dataset<tf.TensorContainer>,
+    valDataset: tf.data.Dataset<tf.TensorContainer>,
+  ): AsyncGenerator<
+    AsyncGenerator<AsyncGenerator<BatchLogs, EpochLogs>, RoundLogs>,
+    void
+  > {
+    if (this.#training !== undefined)
+      throw new Error(
+        "training already running, stop it before launching a new one",
+      );
+
+    try {
+      this.#training = this.#runRounds(dataset, valDataset);
+      yield* this.#training;
+    } finally {
+      this.#training = undefined;
+    }
+  }
+
+  async *#runRounds(
+    dataset: tf.data.Dataset<tf.TensorContainer>,
+    valDataset: tf.data.Dataset<tf.TensorContainer>,
+  ): AsyncGenerator<
+    AsyncGenerator<AsyncGenerator<BatchLogs, EpochLogs>, RoundLogs>,
+    void
+  > {
+    const totalRound = Math.trunc(this.#epochs / this.#roundDuration);
+    let previousRoundWeights: WeightsContainer | undefined;
+    for (let round = 0; round < totalRound; round++) {
+      await this.#client.onRoundBeginCommunication(this.model.weights, round);
+
+      yield this.#runRound(dataset, valDataset);
+
+      let localWeights = this.model.weights;
+      if (this.#privacy !== undefined)
+        localWeights = await applyPrivacy(
+          previousRoundWeights,
+          localWeights,
+          this.#privacy,
+        );
+
+      const networkWeights = await this.#client.onRoundEndCommunication(
+        localWeights,
+        round,
+      );
+
+      if (networkWeights !== undefined) this.model.weights = networkWeights;
```

Comment:
> Is networkWeights expected to be undefined at some point?

~~if the aggregation failed, with a timeout for eg, it'll return `undefined`.~~

ho, you're right, it was a previous idea to return undef on failure but that's not the case anymore, it will return the local weights upon failure, thanks for pointing that out! I just added a comment on `Client.onRoundEndCommunication` stating that.

> If my previous comment on networkWeights and locaWeights being the same when training in local is correct, maybe check if they are already equal too?

correct, the local client simply returns the local weights.
hum, but I don't get the reason behind checking if the weights are already equal, WDYM?

Reply:
I meant to check for equality to not unnecessarily copy weights if they are already the same, but I guess it's only a shallow copy so it does take much time